crates/factor-key-value/src/host.rs

```diff
@@ -141,12 +177,235 @@ impl key_value::HostStore for KeyValueDispatch {
+        &mut self,
+        bucket: Resource<wasi_keyvalue::batch::Bucket>,
+        keys: Vec<String>,
+    ) -> std::result::Result<(), wasi_keyvalue::store::Error> {
+        let store = self.get_store_wasi(bucket)?;
+        store
+            .delete_many(keys.iter().map(|k| k.to_string()).collect())
+            .await
+            .map_err(to_wasi_err)
+    }
+}
+
+#[async_trait]
+impl wasi_keyvalue::atomics::HostCas for KeyValueDispatch {
+    async fn new(
+        &mut self,
+        bucket: Resource<wasi_keyvalue::atomics::Bucket>,
+        key: String,
+    ) -> Result<Resource<wasi_keyvalue::atomics::Cas>, wasi_keyvalue::store::Error> {
+        let bucket_rep = bucket.rep();
+        let bucket: Resource<Bucket> = Resource::new_own(bucket_rep);
+        let store = self.get_store_wasi(bucket)?;
+        let cas = store
+            .new_compare_and_swap(bucket_rep, &key)
+            .await
+            .map_err(to_wasi_err)?;
+        self.compare_and_swaps
+            .push(cas)
+            .map_err(|()| {
+                spin_world::wasi::keyvalue::store::Error::Other(
+                    "too many compare_and_swaps opened".to_string(),
+                )
+            })
+            .map(Resource::new_own)
+    }
+
+    async fn current(
+        &mut self,
+        cas: Resource<wasi_keyvalue::atomics::Cas>,
+    ) -> Result<Option<Vec<u8>>, wasi_keyvalue::store::Error> {
+        let cas = self
+            .get_cas(cas)
+            .map_err(|e| wasi_keyvalue::store::Error::Other(e.to_string()))?;
+        cas.current().await.map_err(to_wasi_err)
+    }
+
+    async fn drop(&mut self, rep: Resource<wasi_keyvalue::atomics::Cas>) -> Result<()> {
+        self.compare_and_swaps.remove(rep.rep());
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl wasi_keyvalue::atomics::Host for KeyValueDispatch {
+    #[instrument(name = "spin_key_value.increment", skip(self, bucket, key, delta), err(level = Level::INFO), fields(otel.kind = "client"))]
+    async fn increment(
+        &mut self,
+        bucket: Resource<wasi_keyvalue::atomics::Bucket>,
+        key: String,
+        delta: i64,
+    ) -> Result<i64, wasi_keyvalue::store::Error> {
+        let store = self.get_store_wasi(bucket)?;
+        store.increment(key, delta).await.map_err(to_wasi_err)
+    }
+
+    #[instrument(name = "spin_key_value.swap", skip(self, cas_res, value), err(level = Level::INFO), fields(otel.kind = "client"))]
+    async fn swap(
+        &mut self,
+        cas_res: Resource<atomics::Cas>,
+        value: Vec<u8>,
+    ) -> Result<std::result::Result<(), CasError>> {
+        let cas_rep = cas_res.rep();
+        let cas = self
+            .get_cas(Resource::<Bucket>::new_own(cas_rep))
+            .map_err(|e| CasError::StoreError(atomics::Error::Other(e.to_string())))?;
+
+        match cas.swap(value).await {
+            Ok(_) => Ok(Ok(())),
+            Err(err) => {
+                if err.to_string().contains("CAS_ERROR") {
```

Comment:
I'm sure you're more conscious than me that this doesn't feel ideal!  But time runs short and we don't want to have to rework the error system, so I guess we go with it for now and re-assess later.  Could you add a "TODO" or "FIXME" or something like that please?

Reply:
Yeah, this was GROSS! I'd was hoping someone would have a better idea than `contains`.

Reply:
@devigned You control the `Cas` trait, I think?  You could give the trait `swap` method a custom error type (such as an enum) that allows for a distinguished "CAS error" case.  (It's internal so can leave as a TODO though!)

Reply:
You are totally right, @itowlson! I'm not sure why that didn't occur to me. I've fixed it by introducing a new error enum.